File: src/main/java/com/oshi/libgenericsettings/presenter/BaseSettingsPresenter.java
package com.oshi.libgenericsettings.presenter;

import android.content.Context;
import android.view.View;

import com.oshi.libgenericsettings.data.BaseViewTypeData;
import com.oshi.libgenericsettings.data.ExpandableTitleCheckableItemsData;
import com.oshi.libgenericsettings.data.ExpandableTitleSimpleItemsData;
import com.oshi.libgenericsettings.data.TitleCheckboxData;
import com.oshi.libgenericsettings.data.TitleIconSeekBarTextData;
import com.oshi.libgenericsettings.data.TitleSecondaryTitleData;
import com.oshi.libgenericsettings.data.TitleSeekBarMinMaxData;
import com.oshi.libgenericsettings.data.TitleSubtitleCheckboxData;
import com.oshi.libgenericsettings.data.IconTitleData;
import com.oshi.libgenericsettings.data.TitleData;
import com.oshi.libgenericsettings.data.TitleSubtitleData;
import com.oshi.libgenericsettings.data.TitleSubtitleExtraCheckboxData;
import com.oshi.libgenericsettings.data.TitleSubtitleExtraData;
import com.oshi.libgenericsettings.data.TitleSubtitleExtraSwitchData;
import com.oshi.libgenericsettings.data.TitleSubtitleSwitchData;
import com.oshi.libgenericsettings.data.TitleSwitchData;
import com.oshi.libgenericsettings.data.TitleUpDownValueData;

import java.util.List;

/**
 * A custom base settings provider which overrides all available types.
 * You should extend this class and override your specific methods.
 */
public class BaseSettingsPresenter implements ISettingsPresenter {

    @Override
    public void onTitleSubtitleSwitchClick(View view, TitleSubtitleSwitchData data, int position) {

    }

    @Override
    public void onTitleSubtitleClick(View view, TitleSubtitleData data, int position) {

    }

    @Override
    public void onTitleSwitchClick(View view, TitleSwitchData data, int position) {

    }

    @Override
    public void onCheckboxTitleSubtitleClick(View view, TitleSubtitleCheckboxData data, int position) {

    }

    @Override
    public void onTitleClick(View view, TitleData data, int position) {

    }

    @Override
    public void onIconTitleClick(View view, IconTitleData data, int position) {

    }

    @Override
    public void onTitleSecondaryTitleClick(View view, TitleSecondaryTitleData data, int position) {

    }

    @Override
    public List<BaseViewTypeData> getItems(Context context) {
        return null;
    }

    @Override
    public void onTitleIconSeekBarTextChanged(View view, TitleIconSeekBarTextData data, int position) {

    }

    @Override
    public void onTitleUpDownValueChanged(View view, TitleUpDownValueData data, int position) {

    }

    @Override
    public void onExpandableSimpleItemClicked(View view, ExpandableTitleSimpleItemsData data, int parentPosition, int subItemPosition) {

    }

    @Override
    public void onTitleCheckboxClick(View view, TitleCheckboxData data, int position) {

    }

    @Override
    public void onTitleSubtitleExtraClick(View view, TitleSubtitleExtraData data, int position) {

    }

    @Override
    public void onTitleSubtitleExtraCheckboxClick(View view, TitleSubtitleExtraCheckboxData data, int position) {

    }

    @Override
    public void onTitleSubtitleExtraSwitchClick(View view, TitleSubtitleExtraSwitchData data, int position) {

    }

    @Override
    public void onTitleSeekBarMinMaxChanged(View view, TitleSeekBarMinMaxData data, int position) {

    }

    @Override
    public void onExpandableCheckableItemClicked(View view, ExpandableTitleCheckableItemsData data, int parentPosition, int subItemPosition) {

    }
}

File: src/main/java/com/oshi/libgenericsettings/presenter/BaseSettingsPresenter.kt
package com.oshi.libgenericsettings.presenter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View
import com.oshi.libgenericsettings.GenericSettings
import com.oshi.libgenericsettings.GenericSettings.get
import com.oshi.libgenericsettings.GenericSettings.set
import com.oshi.libgenericsettings.data.*
import com.oshi.libgenericsettings.helper.GLog

/**
 * A custom base settings provider which overrides all available types.
 * You should extend this class and override your specific methods.
 */
open class BaseSettingsPresenter(val recyclerView : RecyclerView) : ISettingsPresenter {

    override fun onTitleSubtitleSwitchClick(view: View, data: TitleSubtitleSwitchData, position: Int) {
        updateSwitchChange(view, data.key, position)
    }

    override fun onTitleSubtitleClick(view: View, data: TitleSubtitleData, position: Int) {}

    override fun onTitleSwitchClick(view: View, data: TitleSwitchData, position: Int) {
        updateSwitchChange(view, data.key, position)
    }

    override fun onCheckboxTitleSubtitleClick(view: View, data: TitleSubtitleCheckboxData, position: Int) {
        updateCheckboxChange(view, data.key, position)
    }

    override fun onTitleClick(view: View, data: TitleData, position: Int) {}

    override fun onIconTitleClick(view: View, data: IconTitleData, position: Int) {}

    override fun onTitleSecondaryTitleClick(view: View, data: TitleSecondaryTitleData, position: Int) {}

    override fun onTitleIconSeekBarTextChanged(view: View, data: TitleIconSeekBarTextData, position: Int) {}

    override fun onTitleUpDownValueChanged(view: View, data: TitleUpDownValueData, position: Int) {}


    override fun onCheckboxTitleClick(view: View, data: TitleCheckboxData, position: Int) {
        updateCheckboxChange(view, data.key, position)
    }

    override fun onTitleSubtitleExtraClick(view: View, data: TitleSubtitleExtraData, position: Int) {}

    override fun onCheckboxTitleSubtitleExtraClick(view: View, data: TitleSubtitleExtraCheckboxData, position: Int) {
        updateCheckboxChange(view, data.key, position)
    }

    override fun onTitleSubtitleExtraSwitchClick(view: View, data: TitleSubtitleExtraSwitchData, position: Int) {
        updateSwitchChange(view, data.key, position)
    }


    override fun getItems(context: Context): List<BaseViewTypeData> {
        return emptyList()
    }

    override fun onExpandCollapseClicked(position: Int) {
        recyclerView.smoothScrollToPosition(position)
    }

    private fun updateCheckboxChange(view : View, key : String?, position : Int) {
        if (!key.isNullOrBlank()) {
            val prefs = GenericSettings.defaultPrefs(view.context)
            val oldValue = prefs[key!!, false]
            prefs[key] = !oldValue!!
            GLog.d("Updated $key with ${!oldValue}")
            recyclerView.adapter.notifyItemChanged(position)
        }
    }

    private fun updateSwitchChange(view : View, key : String?, position : Int) {
        if (!key.isNullOrBlank()) {
            val prefs = GenericSettings.defaultPrefs(view.context)
            val oldValue = prefs[key!!, false]
            prefs[key] = !oldValue!!
            GLog.d("Updated $key with ${!oldValue}")
            recyclerView.adapter.notifyItemChanged(position)
        }
    }
}